skada/_subspace.py

```diff
@@ -366,3 +373,265 @@ def TransferComponentAnalysis(
+            X_target, self.X_target_
+        ):
+            K = self._get_kernel_matrix(X_source, X_target)
+            X_ = K @ self.A_
+        else:
+            Ks = pairwise_kernels(X, self.X_source_, metric=self.kernel)
+            Kt = pairwise_kernels(X, self.X_target_, metric=self.kernel)
+            K = np.concatenate((Ks, Kt), axis=1)
+            X_ = K @ self.A_
+        return X_
+
+    def _get_mmd_matrix(self, ns, nt, sample_domain):
+        Mss = (1 / (ns**2)) * np.ones((ns, ns))
+        Mtt = (1 / (nt**2)) * np.ones((nt, nt))
+        Mst = -(1 / (ns * nt)) * np.ones((ns, nt))
+        M = np.block([[Mss, Mst], [Mst.T, Mtt]])
+        return M
+
+    def _get_kernel_matrix(self, X_source, X_target):
+        Kss = pairwise_kernels(X_source, metric=self.kernel)
+        Ktt = pairwise_kernels(X_target, metric=self.kernel)
+        Kst = pairwise_kernels(X_source, X_target, metric=self.kernel)
+        K = np.block([[Kss, Kst], [Kst.T, Ktt]])
+        return K
+
+    def fit(self, X, y=None, sample_domain=None, **kwargs):
+        """Fit adaptation parameters.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The source data.
+        y : array-like, shape (n_samples,)
+            The source labels.
+        sample_domain : array-like, shape (n_samples,)
+            The domain labels (same as sample_domain).
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        X, sample_domain = check_X_domain(
+            X,
+            sample_domain,
+            allow_multi_source=True,
+            allow_multi_target=True,
+        )
+        X_source, X_target = source_target_split(X, sample_domain=sample_domain)
+
+        if self.n_components is None:
+            n_components = min(min(X_source.shape), min(X_target.shape))
+        else:
+            n_components = self.n_components
+        self.random_state_ = check_random_state(self.random_state)
+        self.X_source_ = X_source
+        self.X_target_ = X_target
+
+        n = X.shape[0]
+        source_mask = extract_source_indices(sample_domain)
+
+        H = np.identity(n) - 1 / n * np.ones((n, n))
+        K = self._get_kernel_matrix(X_source, X_target)
+        M = self._get_mmd_matrix(X_source.shape[0], X_target.shape[0], sample_domain)
+        M /= np.linalg.norm(M)
+        G = np.identity(n)
+
+        # minimization of the objective function
+        # \min_{A} tr(A^T K M K A) + tradeoff * (||A_s||_{2, 1} + ||A_t||_F^2)
+        # s.t. A^T K H K^T A = I
+        EPS_eigval = 1e-12
+        for i in range(self.max_iter):
+            # update A
+            B = K @ M @ K.T + self.tradeoff * G
+            B = B + EPS_eigval * np.identity(n)
+            C = K @ H @ K.T
+            C = C + EPS_eigval * np.identity(n)
+            solution = scipy.linalg.solve(B, C)
+            phi, A = scipy.linalg.eigh(solution)
```

Comment:
`scipy.linalg.eigh(B, C)` solves `B X = C X D` with `X` and `D` the eigenvectors and eigenvalues respectively. So `phi, A = scipy.linalg.eigh(B, C)`  is better.

Reply:
I don'tw know why and how, but if I don't pass by scipy.linalg.solve (as it was done in TCA), I get really weird (I'm sending you the image) when there is tradeoff, but exactly what we want (the constraint is what we want)
![Figure_5](https://github.com/scikit-adaptation/skada/assets/61913555/e018d9fa-8e85-46b0-84bc-08a7076a1c1b)
What is exactly the difference between using solve and not using it?

Reply:
(don't worry about answering now, I have plenty of other things to do)

Reply:
Solving $B X = C X D$ or $C^{-1} B X = X D$ with respect $X$ and $D$ is the same. Have you checked the eigenvalues of $B$ and $C$ are positive (`np.linalg.eigvalsh(B)` and `np.linalg.eigvalsh(C)`)?

Reply:
Does the loss equal to `0` and are the contraints respected when `tradeoff = 0`?